community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java

```diff
@@ -732,23 +736,23 @@ public boolean hasLabel( Label label )
     @Override
     public Iterable<Label> getLabels()
     {
-        try ( Statement statement = actions.statement() )
-        {
-            PrimitiveIntIterator labels = statement.readOperations().nodeGetLabels( getId() );
-            return asList( map( labelId -> convertToLabel( statement, labelId ), labels ) );
-        }
-        catch ( EntityNotFoundException e )
-        {
-            throw new NotFoundException( "Node not found", e );
-        }
-    }
-
-    private Label convertToLabel( Statement statement, int labelId )
-    {
-        try
+        KernelTransaction transaction = safeAcquireTransaction();
+        try ( Statement ignore = actions.statement();
+              NodeCursor nodes = transaction.cursors().allocateNodeCursor() )
         {
-            return label( statement.readOperations().labelGetName( labelId ) );
-
+            transaction.dataRead().singleNode( nodeId, nodes );
+            if ( !nodes.next() )
+            {
+                throw new NotFoundException( "Node not found" );
+            }
+            LabelSet labelSet = nodes.labels();
+            TokenRead tokenRead = transaction.tokenRead();
+            ArrayList<Label> list = new ArrayList<>( labelSet.numberOfLabels() );
+            for ( int i = 0; i < labelSet.numberOfLabels(); i++ )
+            {
+                list.add( label( tokenRead.labelGetName( labelSet.label( i ) ) ) );
+            }
+            return list;
```

Comment:
Maybe we should make the LabelSet iterable instead, if we are anyway creating that object?

Reply:
That would mean an extra object though? We would need to instantiate the iterator, would probably not be much worse since escape analysis should put it on the stack but still...

Reply:
Eh, not sure I follow... I'm was thinking to return `nodes.labels()` here, which means we allocate the LabelSet, and the iterator in the caller. With the current impl we allocate LabelSet + ArrayList here, and iterator in the caller. I realise that that wasn't really obvious from my previous statement.

Reply:
Oops, forgot about converting to strings... nevermind!